bindings/rust/s2n-tls/src/config.rs

```diff
@@ -291,6 +298,79 @@ impl Builder {
         Ok(self)
     }
 
+    /// Corresponds to [s2n_config_add_cert_chain_and_key_to_store].
+    pub fn load_chain(&mut self, chain: CertificateChain<'static>) -> Result<&mut Self, Error> {
+        // Out of an abudance of caution, we hold a reference to the CertificateChain
+        // regardless of whether add_to_store fails or succeeds. We have limited
+        // visibility into the failure modes, so this behavior ensures that _if_
+        // the C library held the reference despite the failure, it would continue
+        // to be valid memory.
+        let result = unsafe {
+            s2n_config_add_cert_chain_and_key_to_store(
+                self.as_mut_ptr(),
+                // SAFETY: audit of add_to_store shows that the certificate chain
+                // is not mutated. https://github.com/aws/s2n-tls/issues/4140
+                chain.as_ptr() as *mut _,
+            )
+            .into_result()
+        };
+        self.context_mut().application_owned_certs.push(chain);
+        result?;
+
+        Ok(self)
+    }
+
+    /// Corresponds to [s2n_config_set_cert_chain_and_key_defaults].
+    pub fn set_default_chains<T: IntoIterator<Item = CertificateChain<'static>>>(
+        &mut self,
+        chains: T,
+    ) -> Result<&mut Self, Error> {
+        // Must be greater than or equal to S2N_CERT_TYPE_COUNT in s2n_certificate.h,
+        // and small enough that stack allocating an array of this length is safe.
+        const CHAINS_MAX_COUNT: usize = 5;
+
+        let mut chain_arrays: [Option<CertificateChain<'static>>; CHAINS_MAX_COUNT] =
+            [None, None, None, None, None];
+        let mut pointer_array = [std::ptr::null_mut(); CHAINS_MAX_COUNT];
+        let mut cert_chain_count = 0;
+
+        for chain in chains.into_iter() {
+            if cert_chain_count >= CHAINS_MAX_COUNT {
+                // drop the reference counts that we previously had taken
+                return Err(Error::bindings(
+                    ErrorType::UsageError,
+                    "InvalidInput",
+                    "A single default can be specified for RSA, ECDSA, 
+                    and RSA-PSS auth types, but more than 3 certs were supplied",
+                ));
+            }
+
+            // SAFETY: manual inspection of set_defaults shows that certificates
+            // are not mutated. https://github.com/aws/s2n-tls/issues/4140
+            pointer_array[cert_chain_count] = chain.as_ptr() as *mut _;
+            chain_arrays[cert_chain_count] = Some(chain);
+
+            cert_chain_count += 1;
+        }
```

Comment:
Nit: I wonder if it'd be better to do something like:
```
let new_chains = chains.into_iter().take(CHAINS_MAX_COUNT);
if (chains.size() > 0) {
    return Err;
}
```
Then your loop only has to do the array setup.

Reply:
Discussed offline.

We couldn't exactly do this, because `take` consumers the full iterator. But we could do something similar by slightly changing the bounds so that CHAINS_MAX_COUNT is strict greater than S2N_CERT_TYPE_COUNT.

So if they pass in 10 certs, it doesn’t really matter that we only pass in 4 to the library, because s2n-tls (C library) will return an error because 4 > 3

However this relies on the Rust/C constants having a very specific, odd relationships, so the current approach is preferred.